File: sdk/python/tests/integrations/ag_ui/_vectors.py
"""Shared deterministic input vectors for the TA suite — mirror of the
TS ``tests/_support/vectors.ts`` role. Pure data, no clocks, no RNG."""

from __future__ import annotations

from typing import Any, Callable

from spendguard.integrations.ag_ui import (
    BudgetSnapshotInput,
    DecisionDeniedInput,
    ReservationCommittedInput,
    ReservationCreatedInput,
    ReservationReleasedInput,
    build_budget_snapshot,
    build_decision_denied,
    build_reservation_committed,
    build_reservation_created,
    build_reservation_released,
)

SNAPSHOT_KW: dict[str, Any] = {
    "budget_id": "budget-dev-monthly",
    "window_instance_id": "0197a001-0000-7000-8000-00000000win1",
    "unit": "usd_micros",
    "unit_id": "0197a001-2222-7000-8000-0000000unit1",
    "remaining_atomic": "24000000",
    "reserved_atomic": "1000000",
    "spent_atomic": "950000",
    "as_of": "2026-06-10T08:00:05Z",
}

CREATED_KW: dict[str, Any] = {
    "decision_id": "0197a001-3333-7000-8000-000000000dec",
    "reservation_id": "0197a001-4444-7000-8000-000000000res",
    "budget_id": "budget-dev-monthly",
    "window_instance_id": "0197a001-0000-7000-8000-00000000win1",
    "unit": "usd_micros",
    "unit_id": "0197a001-2222-7000-8000-0000000unit1",
    "amount_atomic_reserved": "1000000",
    "decision": "ALLOW",
    "ttl_expires_at": "2026-06-10T08:05:00Z",
    "reason_codes": ["within_budget"],
    "matched_rule_ids": ["rule-default-allow"],
    "run_id": "0197a001-5555-7000-8000-000000000run",
    "llm_call_id": "0197a001-6666-7000-8000-00000000call",
    "event_time": "2026-06-10T08:00:00Z",
}

COMMITTED_KW: dict[str, Any] = {
    "decision_id": "0197a001-3333-7000-8000-000000000dec",
    "reservation_id": "0197a001-4444-7000-8000-000000000res",
    "budget_id": "budget-dev-monthly",
    "window_instance_id": "0197a001-0000-7000-8000-00000000win1",
    "unit": "usd_micros",
    "unit_id": "0197a001-2222-7000-8000-0000000unit1",
    "amount_atomic_estimated": "950000",
    "outcome": "SUCCESS",
    "run_id": "0197a001-5555-7000-8000-000000000run",
    "llm_call_id": "0197a001-6666-7000-8000-00000000call",
    "event_time": "2026-06-10T08:00:03Z",
}

RELEASED_KW: dict[str, Any] = {
    "reservation_id": "0197a001-4444-7000-8000-000000000res",
    "decision_id": "0197a001-3333-7000-8000-000000000dec",
    "reason_codes": ["client_timeout"],
    "ledger_transaction_id": "0197a001-7777-7000-8000-00000000ltx1",
    "run_id": "0197a001-5555-7000-8000-000000000run",
    "llm_call_id": "0197a001-6666-7000-8000-00000000call",
    "event_time": "2026-06-10T08:00:04Z",
}

DENIED_KW: dict[str, Any] = {
    "decision_id": "0197a001-8888-7000-8000-000000000dec",
    "denied_kind": "DENY",
    "reason_codes": ["BUDGET_EXCEEDED"],
    "matched_rule_ids": ["rule-hard-cap"],
    "budget_id": "budget-dev-monthly",
    "window_instance_id": "0197a001-0000-7000-8000-00000000win1",
    "unit": "usd_micros",
    "unit_id": "0197a001-2222-7000-8000-0000000unit1",
    "run_id": "0197a001-5555-7000-8000-000000000run",
    "llm_call_id": "0197a001-6666-7000-8000-00000000call",
    "event_time": "2026-06-10T08:00:06Z",
}

# (builder, input class, kwargs, locked §5.2 name) — table order.
ALL_BUILDERS: list[tuple[Callable[..., dict[str, Any]], type, dict[str, Any], str]] = [
    (build_budget_snapshot, BudgetSnapshotInput, SNAPSHOT_KW,
     "spendguard.budget.snapshot"),
    (build_reservation_created, ReservationCreatedInput, CREATED_KW,
     "spendguard.reservation.created"),
    (build_reservation_committed, ReservationCommittedInput, COMMITTED_KW,
     "spendguard.reservation.committed"),
    (build_reservation_released, ReservationReleasedInput, RELEASED_KW,
     "spendguard.reservation.released"),
    (build_decision_denied, DecisionDeniedInput, DENIED_KW,
     "spendguard.decision.denied"),
]


def build_all(timestamp_ms: int | None = None) -> list[dict[str, Any]]:
    """One built event per builder, from the canonical vectors."""
    return [
        fn(cls(**kw), timestamp_ms=timestamp_ms)
        for fn, cls, kw, _name in ALL_BUILDERS
    ]

File: sdk/typescript-ag-ui/tests/_support/vectors.ts
// Shared deterministic input vectors for the @spendguard/ag-ui test suite.
//
// Every value is a fixed literal — no clocks, no RNG (design.md §11.3).
// IDs are substrate-shaped inputs (UUIDv7-style), never minted here
// (design.md §11.6).
import type {
  BudgetSnapshotInput,
  DecisionDeniedInput,
  ReservationCommittedInput,
  ReservationCreatedInput,
  ReservationReleasedInput,
} from "../../src/events.js";

export const TS_MS = 1765843200000;

export const SNAPSHOT_MIN: BudgetSnapshotInput = {
  budgetId: "budget-dev-monthly",
  windowInstanceId: "0197a001-0000-7000-8000-00000000win1",
  unit: "usd_micros",
  remainingAtomic: "25000000",
  reservedAtomic: "0",
  spentAtomic: "0",
  asOf: "2026-06-10T07:59:00Z",
};

export const SNAPSHOT_MAX: BudgetSnapshotInput = {
  ...SNAPSHOT_MIN,
  unitId: "0197a001-2222-7000-8000-0000000unit1",
};

export const CREATED_MIN: ReservationCreatedInput = {
  decisionId: "0197a001-aaaa-7000-8000-000000000d01",
  reservationId: "0197a001-bbbb-7000-8000-000000000r01",
  budgetId: "budget-dev-monthly",
  windowInstanceId: "0197a001-0000-7000-8000-00000000win1",
  unit: "usd_micros",
  amountAtomicReserved: "1000000",
  decision: "ALLOW",
  ttlExpiresAt: "2026-06-10T08:00:00Z",
  eventTime: "2026-06-10T07:59:58Z",
};

export const CREATED_MAX: ReservationCreatedInput = {
  ...CREATED_MIN,
  unitId: "0197a001-2222-7000-8000-0000000unit1",
  decision: "ALLOW_WITH_CAPS",
  reasonCodes: ["degrade_applied", "model_capped"],
  matchedRuleIds: ["rule-cap-model", "rule-budget-soft"],
  runId: "0197a001-cccc-7000-8000-00000000run1",
  llmCallId: "0197a001-dddd-7000-8000-0000000call1",
};

export const COMMITTED_MIN: ReservationCommittedInput = {
  decisionId: "0197a001-aaaa-7000-8000-000000000d01",
  reservationId: "0197a001-bbbb-7000-8000-000000000r01",
  budgetId: "budget-dev-monthly",
  windowInstanceId: "0197a001-0000-7000-8000-00000000win1",
  unit: "usd_micros",
  amountAtomicEstimated: "950000",
  outcome: "SUCCESS",
  eventTime: "2026-06-10T08:00:02Z",
};

export const COMMITTED_MAX: ReservationCommittedInput = {
  ...COMMITTED_MIN,
  unitId: "0197a001-2222-7000-8000-0000000unit1",
  amountAtomicObserved: "940123",
  runId: "0197a001-cccc-7000-8000-00000000run1",
  llmCallId: "0197a001-dddd-7000-8000-0000000call1",
};

export const RELEASED_MIN: ReservationReleasedInput = {
  reservationId: "0197a001-bbbb-7000-8000-000000000r01",
  reasonCodes: ["client_timeout"],
  eventTime: "2026-06-10T08:00:30Z",
};

export const RELEASED_MAX: ReservationReleasedInput = {
  ...RELEASED_MIN,
  decisionId: "0197a001-aaaa-7000-8000-000000000d01",
  reasonCodes: ["provider_error", "run_cancelled"],
  ledgerTransactionId: "0197a001-eeee-7000-8000-000000000tx1",
  runId: "0197a001-cccc-7000-8000-00000000run1",
  llmCallId: "0197a001-dddd-7000-8000-0000000call1",
};

export const DENIED_MIN: DecisionDeniedInput = {
  decisionId: "0197a001-ffff-7000-8000-000000000d02",
  deniedKind: "DENY",
  reasonCodes: ["BUDGET_EXHAUSTED"],
  eventTime: "2026-06-10T08:01:00Z",
};

export const DENIED_MAX: DecisionDeniedInput = {
  ...DENIED_MIN,
  deniedKind: "STOP_RUN_PROJECTION",
  reasonCodes: ["RUN_BUDGET_PROJECTION_EXCEEDED", "RUN_CEILING"],
  matchedRuleIds: ["rule-run-projection"],
  budgetId: "budget-dev-monthly",
  windowInstanceId: "0197a001-0000-7000-8000-00000000win1",
  unit: "usd_micros",
  unitId: "0197a001-2222-7000-8000-0000000unit1",
  runId: "0197a001-cccc-7000-8000-00000000run1",
  llmCallId: "0197a001-dddd-7000-8000-0000000call1",
};

/** Recursively freeze an object graph (TP-12 input-mutation guard). */
export function deepFreeze<T>(obj: T): T {
  if (obj !== null && typeof obj === "object") {
    for (const v of Object.values(obj)) {
      deepFreeze(v);
    }
    Object.freeze(obj);
  }
  return obj;
}
